### ISSUE TITLE
AdGuard adblocker no longer detected

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#2794]
## Bug Description

The popular [AdGuard ad blocker](https://chrome.google.com/webstore/detail/adguard-adblocker/bgnkhhnnamicmpeenaelnjfhikgbkllg) is no longer detected by our detection mechanism.

This allows users to proceed with AdSense setup where it will fail when AdSense module API requests are blocked.

![image](https://user-images.githubusercontent.com/1621608/107667509-4dbd1c00-6c98-11eb-8e7a-e953e784e82f.png)

![image](https://user-images.githubusercontent.com/1621608/107667515-4eee4900-6c98-11eb-8ce7-0d02c501875d.png)

![image](https://user-images.githubusercontent.com/1621608/107667527-5150a300-6c98-11eb-9ff6-9809de6ac41d.png)

Another popular ad blocker, Ad Block Plus is still detected so the breakage seems to be a result of changes to the blocking rules no longer preventing our test script from loading.

![image](https://user-images.githubusercontent.com/1621608/107667782-9aa0f280-6c98-11eb-8e17-4938999869af.png)

The chrome extension alone reports 10M+ users and their website advertises 30M+ cross-platform users so this will likely impact some of our users.

## Steps to reproduce

1. Start with the AdSense module inactive
1. Install [AdGuard ad blocker for your browser](https://github.com/AdguardTeam/AdguardBrowserExtension#installation)
2. Go to Site Kit settings > Connect tab
3. Notice there is no ad blocker detected warning
4. Activate AdSense and start setup
5. Note errors preventing setup from completing

## Additional Context

This is likely due to a change in the AdGuard pattern matching rather than a regression in Site Kit itself since the detection logic is purely based on the URL of the [detection script](https://github.com/google/site-kit-wp/blob/5817746473cc66880feb5c0b621ce2b3cddf0c44/assets/js/pagead2.ads.js).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The ad blocker detection script should be updated so that it is blocked by the current [AdGuard](https://adguard.com/) ad blocker
* Other currently supported ad blockers should continue to be detected

## Implementation Brief

* Remove dedicated external ad-detection script entirely
	- Entry in `Assets`
	- References to `googlesitekit-pagead2.ads` handle
	- Entry point in webpack.config
* Add [`just-detect-adblock`](https://www.npmjs.com/package/just-detect-adblock) as a dependency
* Update `resolvers.isAdBlockerActive` in `assets/js/modules/adsense/datastore/adblocker.js` to receive the value based on the library-based detection function (adding a new control to return the value asynchronously)
* Update the `checkRequirements` function for AdSense to asynchronously check `isAdBlockerActive`
* **Merge https://github.com/google/site-kit-wp/pull/2882**

### Test Coverage

* Update tests for adblocker selector in `datastore/adblocker.test.js`

### Visual Regression Changes

* None expected

## QA Brief

* Navigate to Settings > connect tab with AdSense inactive
* Reload the page with different adblocker extensions active (including AdGuard, AdBlock, uBlock Origin)
* Ensure AdSense displays the adblocker warning and prevents setup/activation

## Changelog entry

* Fix bug where AdGuard and uBlock ad blockers were no longer being recognized by Site Kit by implementing a more holistic solution for ad blocker detection relying on the `just-detect-adblock` library.

**Comment by aaemnnosttv** [comment]
@felixarntz – another user [recently reported](https://github.com/google/site-kit-wp/issues/1666#issuecomment-783699862) that [uBlock Origin](https://github.com/gorhill/uBlock) was not detected either. The [Chrome extension](https://chrome.google.com/webstore/detail/ublock-origin/cjpalhdlnbpafiamejdnhcphjbkeiagm?hl=en-US) has 10M+ users.

**Comment by felixarntz** [comment]
Awesome to improve this holistically - IB ✅

**Comment by wpdarren** [comment]
### QA Update: Engineer to Confirm ⚠️ 

@aaemnnosttv @felixarntz  

This passes the QAB, but I decided to try break this and noticed if I am a new Site Kit user (or disconnect and reset data) when I am directed to the Site Kit Dashboard, the warning on Google Adsense does appear but, the `connect service` link appears.  

![image](https://user-images.githubusercontent.com/73545194/109973094-f3612980-7d32-11eb-9cd0-80d3f7f153b5.png)

When I click on the link I am taken to a page (after connecting my Google account) where it says that an Ad blocker is detected and stops you from connecting.
 
![image](https://user-images.githubusercontent.com/73545194/109973143-ff4ceb80-7d32-11eb-85f0-8e0465c0080a.png)

When you click `cancel ` you are redirected to the settings > connection page and another message appears. 

![image](https://user-images.githubusercontent.com/73545194/109973927-f6104e80-7d33-11eb-8a53-43c4b30b4654.png)

In my opinion, it would be a better UX, if the `connect service` link was hidden.

**Verified:**

While AdSense is inactive, and the page with adblocker extension is added, there's a warning that prevents you from setup/activation from the settings > connection screen - [Screenshot](https://user-images.githubusercontent.com/73545194/109973628-9d40b600-7d33-11eb-8342-1ead8f65dbae.png)

**Comment by aaemnnosttv** [comment]
@wpdarren regarding the UX issue you found, this may have been a regression at some point 🤔  
It's a good question but it shouldn't block this issue since it's not related to the changes here. Let's open that as a new issue 👍 

We don't have a specific list of adblockers that we support but all the major ones like AdBlock (Plus), AdGuard, uBlock Origin are the ones that come to mind for me.

**Comment by wpdarren** [comment]
### QA Update:  Fail ❌ 

@aaemnnosttv With uBlock Origin Chrome extension activated, I am able to connect Adsense from within the settings. The adblocker warning does not appear. 

![image](https://user-images.githubusercontent.com/73545194/110124284-dabd4600-7dfc-11eb-82e9-33a10b03266d.png)

**Verified:**

AdBlock (Plus), AdGuard:  While AdSense is inactive, and the page with adblocker extension is added, there's a warning that prevents you from setup/activation from the settings > connection screen 

![image](https://user-images.githubusercontent.com/73545194/110123880-6b475680-7dfc-11eb-86b6-82a7cc74fc0f.png)

**Comment by aaemnnosttv** [comment]
Thanks @wpdarren – I've opened #2925 which adds another layer of adblocker detection that properly detects uBlock (and probably many others).

Assigning to @felixarntz to review the follow-up.

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

**Verified:**

AdBlock (Plus), AdGuard: While AdSense is inactive, and the page with adblocker extension is added, there's a warning that prevents you from setup/activation from the settings > connection screen

![image](https://user-images.githubusercontent.com/73545194/110123880-6b475680-7dfc-11eb-86b6-82a7cc74fc0f.png)
